subsys/usb/host/usbh_shell.c

```diff
@@ -0,0 +1,465 @@
+	shell_print(shell, "bNumConfigurations\t\t%u", desc->bNumConfigurations);
+}
+
+static int bazfoo_request(struct usbh_contex *const ctx,
+			  struct uhc_transfer *const xfer,
+			  int err)
+{
+	const struct device *dev = ctx->dev;
+
+	shell_print(ctx_shell, "xfer finished %p, err %d", xfer, err);
+
+	while (!k_fifo_is_empty(&xfer->done)) {
+		struct net_buf *buf;
+
+		buf = net_buf_get(&xfer->done, K_NO_WAIT);
+		if (buf) {
+			if (xfer->ep == USB_CONTROL_EP_IN &&
+			    buf->len == sizeof(struct usb_device_descriptor)) {
+				print_dev_desc(ctx_shell, (void *)buf->data);
+			} else {
+				shell_hexdump(ctx_shell, buf->data, buf->len);
+			}
+
+			uhc_xfer_buf_free(dev, buf);
+		}
+	}
+
+	return uhc_xfer_free(dev, xfer);
+}
+
+static int bazfoo_connected(struct usbh_contex *const uhs_ctx)
+{
+	shell_print(ctx_shell, "USB device connected");
+
+	return 0;
+}
+
+static int bazfoo_removed(struct usbh_contex *const uhs_ctx)
+{
+	shell_print(ctx_shell, "USB device removed");
+
+	return 0;
+}
+
+static int bazfoo_rwup(struct usbh_contex *const uhs_ctx)
+{
+	shell_print(ctx_shell, "Bus remote wakeup event");
+
+	return 0;
+}
+
+static int bazfoo_suspended(struct usbh_contex *const uhs_ctx)
+{
+	shell_print(ctx_shell, "Bus suspended");
+
+	return 0;
+}
+
+static int bazfoo_resumed(struct usbh_contex *const uhs_ctx)
+{
+	shell_print(ctx_shell, "Bus resumed");
+
+	return 0;
+}
+
+USBH_DEFINE_CLASS(bazfoo) = {
+	.request = bazfoo_request,
+	.connected = bazfoo_connected,
+	.removed = bazfoo_removed,
+	.rwup = bazfoo_rwup,
+	.suspended = bazfoo_suspended,
+	.resumed = bazfoo_resumed,
+};
+
+static int cmd_desc_device(const struct shell *sh,
+			   size_t argc, char **argv)
+{
+	uint8_t addr;
+	int err;
+
+	addr = strtol(argv[1], NULL, 16);
+
+	err = usbh_req_desc_dev(uhc_dev, addr);
+	if (err) {
+		shell_print(sh, "Failed to request device descriptor");
+	}
+
+	return err;
+}
+
+static int cmd_desc_config(const struct shell *sh,
+			   size_t argc, char **argv)
+{
+	uint8_t addr;
+	uint8_t cfg;
+	int err;
+
+	addr = strtol(argv[1], NULL, 16);
+	cfg = strtol(argv[2], NULL, 10);
+
+	/* TODO: cfg is ignored, add to usbh_req_desc_cfg */
+	err = usbh_req_desc_cfg(uhc_dev, addr, cfg, 128);
+	if (err) {
+		shell_print(sh, "Failed to request configuration descriptor");
+	}
+
+	return err;
+}
+
+static int cmd_desc_string(const struct shell *sh,
+			   size_t argc, char **argv)
+{
+	const uint8_t type = USB_DESC_STRING;
+	uint8_t addr;
+	uint8_t id;
+	uint8_t idx;
+	int err;
+
+	addr = strtol(argv[1], NULL, 16);
+	id = strtol(argv[2], NULL, 10);
+	idx = strtol(argv[3], NULL, 10);
+
+	err = usbh_req_desc(uhc_dev, addr, type, idx, id, 128);
+	if (err) {
+		shell_print(sh, "Failed to request configuration descriptor");
+	}
+
+	return err;
+}
+
+static int cmd_feature_set_halt(const struct shell *sh,
+				size_t argc, char **argv)
+{
+	uint8_t addr;
+	uint8_t ep;
+	int err;
+
+	addr = strtol(argv[1], NULL, 16);
+	ep = strtol(argv[2], NULL, 16);
```

Comment:
Same as with address, why base16?

Reply:
Because we print/display endpoint address always base16.